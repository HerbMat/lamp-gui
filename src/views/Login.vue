<template>
  <div class="flex-box center">
    <div class="container">
      <form class="login-form-layout">
        <app-errors v-bind:errors="errors" />
        <label for="username">Username</label>
        <input type="text" id="username" v-model="user.username"/>
        <label for="password">Password</label>
        <input type="password" id="password" v-model="user.password"/>
        <input type="submit" id="login-submit" @click.prevent="login" value="Log in"/>
      </form>
    </div>
  </div>
</template>

<script>
import router from '../router'
import FormErrors from '../components/FormErrors'

export default {
  name: 'Login',
  data: () => {
    return {
      user: {
        username: '',
        password: ''
      },
      errors: []
    }
  },
  methods: {
    login () {
      if (this.checkForm()) {
        this.$store.dispatch('login', {
          username: this.user.username,
          password: this.user.password
        }).then(() => router.replace('/'))
          .catch(error => {
            this.errors = []
            if (error.response.status === 401) {
              this.errors.push('Bad username or password')
            } else {
              this.errors.push('There was some kind of error')
            }
          })
      }
    },
    checkForm () {
      this.errors = []
      if (!this.user.username) {
        this.errors.push('Name required.')
      }
      if (!this.user.password) {
        this.errors.push('Password required.')
      }
      return this.errors.length === 0
    }
  },
  components: {
    'app-errors': FormErrors
  }
}
</script>

<style lang="scss" scoped>
  .login-form-layout {
    margin-left: 20%;
    margin-right: 20%;
    min-width: 3rem;
    display: grid;
    grid-template-columns: repeat(1, 1fr);
    grid-gap: 1rem;
  }

  .login-form-layout > input {
    width: 100%;
    min-width: 3rem;
    height: 3rem;
  }

  .login-form-layout > label {
    display: block;
    text-align: center;
  }
</style>
