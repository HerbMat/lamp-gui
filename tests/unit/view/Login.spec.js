import Vuex from 'vuex'
import { mount, shallowMount } from '@vue/test-utils'
import Login from '../../../src/views/Login'

// const localVue = createLocalVue()
// localVue.use(Vuex)

describe('Login.vue', () => {
  const wrapper = mount(Login)

  it('has data', () => {
    expect(typeof Login.data).toBe('function')
    expect(Login.data().user.username).toBe('')
    expect(Login.data().user.password).toBe('')
    expect(Login.data().errors).toEqual([])
  })

  it('Snapshot match', () => {
    const wrapperToCheck = shallowMount(Login, {})
    expect(wrapperToCheck).toMatchSnapshot()
  })

  it('Prints form errors when from empty', () => {
    wrapper.find('#login-submit').trigger('click')
    expect(wrapper.vm.errors.length).toEqual(2)
    expect(wrapper.find('.error').exists()).toBeTruthy()
  })
})
