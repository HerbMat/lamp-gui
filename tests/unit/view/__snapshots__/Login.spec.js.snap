// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue Snapshot match 1`] = `
<div class="flex-box center">
  <div class="container">
    <form class="login-form-layout">
      <app-errors-stub errors=""></app-errors-stub> <label for="username">Username</label> <input type="text" id="username"> <label for="password">Password</label> <input type="password" id="password"> <input type="submit" id="login-submit" value="Log in">
    </form>
  </div>
</div>
`;
